File: docs/mp2i/files/Snake/snake.c
#include <ncurses.h>
#include <stdlib.h>
#include <time.h>

// Structure pour une position
typedef struct Position {
    int x;
    int y;
} Position;

// Noeud de la liste chaînée
typedef struct Node {
    Position pos;
    struct Node *next;
} Node;


// Liste chaînée du serpent
typedef struct Snake {
    Node *head;  // Tête du serpent (premier maillon)
    Node *tail;  // Queue du serpent (dernier maillon)
    int length;
} Snake;

// Directions
typedef enum {
    UP,
    DOWN,
    LEFT,
    RIGHT
} Direction;

// Fonctions de gestion de la liste chaînée
Snake* snake_init(int x, int y) {
    Snake *snake = malloc(sizeof(Snake));
    Node *node = malloc(sizeof(Node));
    node->pos.x = x;
    node->pos.y = y;
    node->next = NULL;
    snake->head = node;
    snake->tail = node;
    snake->length = 1;
    return snake;
}

// Ajouter après la queue - O(1)
void snake_add_tail(Snake *snake, int x, int y) {
    Node *new_node = malloc(sizeof(Node));
    new_node->pos.x = x;
    new_node->pos.y = y;
    new_node->next = NULL;
    snake->tail->next = new_node;
    snake->tail = new_node;
    snake->length++;
}

// Retirer depuis la tête - O(1)
void snake_remove_head(Snake *snake) {
    if (snake->length <= 1) return;
    
    Node *old_head = snake->head;
    snake->head = snake->head->next;
    free(old_head);
    snake->length--;
}

// Vérifier collision avec le serpent
int snake_collides_with_self(Snake *snake, int x, int y) {
    Node *current = snake->head;
    while (current != NULL) {
        if (current->pos.x == x && current->pos.y == y) {
            return 1;
        }
        current = current->next;
    }
    return 0;
}

// Dessiner le serpent
void snake_draw(Snake *snake) {
    Node *current = snake->head;
    while (current != NULL) {
        // La queue (tail) est la tête du serpent
        mvaddch(current->pos.y, current->pos.x, 
                (current == snake->tail) ? 'O' : 'o');
        current = current->next;
    }
}

// Libérer la mémoire
void snake_free(Snake *snake) {
    Node *current = snake->head;
    while (current != NULL) {
        Node *next = current->next;
        free(current);
        current = next;
    }
    free(snake);
}

// Générer une pomme aléatoire
Position generate_apple(Snake *snake, int max_x, int max_y) {
    Position apple;
    do {
        apple.x = rand() % (max_x - 2) + 1;
        apple.y = rand() % (max_y - 2) + 1;
    } while (snake_collides_with_self(snake, apple.x, apple.y));
    return apple;
}

// Initialiser le terminal
void terminal_init() {
    initscr();              // Initialiser ncurses
    cbreak();               // Désactiver la bufferisation de ligne
    noecho();               // Ne pas afficher les touches tapées
    keypad(stdscr, TRUE);   // Activer les touches spéciales (flèches)
    nodelay(stdscr, TRUE);  // Rendre getch() non-bloquant
    curs_set(0);            // Cacher le curseur
}

// Nettoyer et restaurer le terminal
void terminal_cleanup() {
    endwin();               // Restaurer le terminal à son état normal
}

int main() {
    // Initialisation
    srand(time(NULL));
    terminal_init();
    
    int max_y, max_x;
    getmaxyx(stdscr, max_y, max_x);
    
    // Initialiser le serpent au centre
    Snake *snake = snake_init(max_x / 2, max_y / 2);
    Direction dir = RIGHT;
    Direction next_dir = RIGHT;
    
    // Générer la première pomme
    Position apple = generate_apple(snake, max_x, max_y);
    
    int game_over = 0;
    int score = 0;
    
    while (!game_over) {
        clear();
        
        // Dessiner les bordures
        for (int i = 0; i < max_x; i++) {
            mvaddch(0, i, '#');
            mvaddch(max_y - 1, i, '#');
        }
        for (int i = 0; i < max_y; i++) {
            mvaddch(i, 0, '#');
            mvaddch(i, max_x - 1, '#');
        }
        
        // Afficher le score
        mvprintw(0, 2, " Score: %d ", score);
        
        // Dessiner la pomme
        mvaddch(apple.y, apple.x, '*');
        
        // Dessiner le serpent
        snake_draw(snake);
        
        refresh();
        
        // Gérer l'entrée clavier
        int ch = getch();
        if (ch == 'q' || ch == 'Q') break;
        
        switch (ch) {
            case KEY_UP:
                if (dir != DOWN) next_dir = UP;
                break;
            case KEY_DOWN:
                if (dir != UP) next_dir = DOWN;
                break;
            case KEY_LEFT:
                if (dir != RIGHT) next_dir = LEFT;
                break;
            case KEY_RIGHT:
                if (dir != LEFT) next_dir = RIGHT;
                break;
        }
        
        dir = next_dir;
        
        // Calculer la nouvelle position de la tête
        int new_x = snake->head->pos.x;
        int new_y = snake->head->pos.y;
        
        switch (dir) {
            case UP:    new_y--; break;
            case DOWN:  new_y++; break;
            case LEFT:  new_x--; break;
            case RIGHT: new_x++; break;
        }
        
        // Vérifier collision avec les murs
        if (new_x <= 0 || new_x >= max_x - 1 || 
            new_y <= 0 || new_y >= max_y - 1) {
            game_over = 1;
            continue;
        }
        
        // Vérifier collision avec soi-même
        if (snake_collides_with_self(snake, new_x, new_y)) {
            game_over = 1;
            continue;
        }
        
        // Ajouter la nouvelle position en queue
        snake_add_tail(snake, new_x, new_y);
        
        // Vérifier si on mange la pomme
        if (new_x == apple.x && new_y == apple.y) {
            score++;
            apple = generate_apple(snake, max_x, max_y);
        } else {
            // Retirer la tête si pas de pomme mangée
            snake_remove_head(snake);
        }
        
        usleep(100000); // 100ms de délai
    }
    
    // Game Over
    clear();
    mvprintw(max_y / 2, max_x / 2 - 10, "GAME OVER!");
    mvprintw(max_y / 2 + 1, max_x / 2 - 10, "Score: %d", score);
    mvprintw(max_y / 2 + 2, max_x / 2 - 15, "Appuyez sur une touche...");
    refresh();
    nodelay(stdscr, FALSE);
    getch();
    
    // Nettoyage
    snake_free(snake);
    terminal_cleanup();
    
    return 0;
}

/* 
 * Compilation:
 * gcc -o snake snake.c -lncurses
 * 
 * Exécution:
 * ./snake
 * 
 * Contrôles:
 * - Flèches directionnelles pour diriger le serpent
 * - 'q' pour quitter
 */

File: docs/itc/files/C7/snake.py
import curses
from random import randint
from time import sleep

#--8<-- [start:car]
SNAKE_HEAD = "O"
SNAKE_BODY = "o"
BORDER = "+"
EMPTY = ' '
FOOD = "*"
#--8<-- [end:car]

RIGHT, UP, LEFT, DOWN = 0, 1, 2, 3
GROW, COLLISION = 1, 2
DIR = {RIGHT : (0, 1), UP : (-1,0), LEFT : (0,-1), DOWN : (1,0)}

def draw_border(maxl, maxc):
    for c in range(0,maxc):
        terminal.addch(0,c,BORDER)
        terminal.addch(maxl,c,BORDER)
    for l in range(0,maxl):
        terminal.addch(l,0,BORDER)
        terminal.addch(l,maxc,BORDER)
    terminal.refresh()

def est_dans(x, lst):
    for elt in lst:
        if elt==x:
            return True
    return False

def init_snake(maxl, maxc, size):
    ls = maxl//2
    sn = []
    for c in range(maxc//2-size//2,maxc//2+size//2):
        sn.append((ls,c))
        terminal.addch(ls,c,SNAKE_BODY)
    sn.append((ls,maxc//2+size//2))
    terminal.addch(ls,maxc//2+size//2,SNAKE_BODY)
    terminal.refresh()
    return sn


def move_snake(pos_snake, dir_snake, maxl, maxc, fl, fc):
    # Ajouter la nouvelle position devant la tête
    hl, hc = pos_snake[-1]
    dl, dc = DIR[dir_snake]
    if hl+dl<1 or hl+dl>maxl or hc+dc<1 or hc+dc>maxc or est_dans((hl+dl,hc+dc),pos_snake):
        return COLLISION
    pos_snake.append((hl+dl,hc+dc))
    terminal.addch(hl,hc,SNAKE_BODY)
    terminal.addch(hl+dl,hc+dc,SNAKE_HEAD)
    terminal.refresh()
    if hl+dl==fl and hc+dc==fc:
        return GROW
    tl, tc = pos_snake.pop(0)
    terminal.addch(tl,tc,EMPTY)
    terminal.refresh()

def make_food(maxl, maxc, snake_pos):
    fl = randint(1,maxl-1)
    fc = randint(1,maxc-1)
    while est_dans((fl,fc),snake_pos):
        fl = randint(1,maxl-1)
        fc = randint(1,maxc-1)
    terminal.addch(fl,fc,FOOD)
    terminal.refresh()
    return fl,fc

def snake(terminal):
    #Vitesse du jeu
    delay = 0.2
    #Initialiser l'écran
    curses.cbreak()              #Désactiver la bufferisation de ligne
    curses.curs_set(0)           #Cacher le curseur
    curses.noecho()               #Ne pas afficher les touches tapées
    terminal.nodelay(1) #Rendre getch() non-bloquant
    terminal.keypad(True)
    terminal.clear()

    # Récupérer les dimensions du terminal
    maxl, maxc = terminal.getmaxyx()
    maxl = maxl-1
    maxc =  maxc - 1
    draw_border(maxl,maxc)
    pos_snake = init_snake(maxl, maxc, 5)
    dir_snake = RIGHT
    running = True
    fl,fc = make_food(maxl,maxc,pos_snake)
    while (running):
        event = move_snake(pos_snake,dir_snake, maxl, maxc, fl, fc)
        if event==GROW:
            fl,fc = make_food(maxl,maxc,pos_snake)
            delay = 0.9 * delay
        if event==COLLISION:
            running = False
        sleep(delay)
        kpress = terminal.getch()
        if (kpress == curses.KEY_UP):
            dir_snake = UP
        elif (kpress == curses.KEY_DOWN):
            dir_snake = DOWN
        elif (kpress == curses.KEY_LEFT):
            dir_snake =  LEFT
        elif (kpress == curses.KEY_RIGHT):
            dir_snake =  RIGHT
        


terminal = curses.initscr()
curses.wrapper(snake)
